### PR TITLE
Implement Misc & Logging API

### LINKED ISSUE #169: Implement Logging & Abort handler API

### PR DESCRIPTION
**Pull request by ablax**
- [x] Misc API functions
- [x] Log API Functions
- [x] Tests
 
Closes #169

### CODE REVIEW

File: src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java [hunk 1]
```diff
@@ -1,67 +1,163 @@
...
+
+        Module module = new Module(wasmBlob);
+        Runtime runtime = new Runtime(module, RuntimeBuilder.DEFAULT_HEAP_PAGES);
+        Memory memory = runtime.getInstance().exports.getMemory("memory");
 
-        log.log(getLogLevel(level), new String(target) + ": " + new String(message));
+        Object[] response = runtime.call("Core_version");
+
+        if (response == null || response[0] == null) {
+            throw new RuntimeException(CORE_VERSION_CALL_FAILED);
```
Comment (stoyan-lime) [inline src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java]: `None` should be returned when the call fails.

File: src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java [hunk 2]
```diff
@@ -70,4 +166,44 @@ private static Level getLogLevel(int i) {
...
+     * Aborts the execution of the runtime with a given message.
+     *
+     * @param messagePtr a pointer-size to the UTF-8 encoded message.
+     */
+    public void extPanicHandlerAbortOnPanicVersion1(RuntimePointerSize messagePtr) {
+        byte[] data = hostApi.getDataFromMemory(messagePtr);
+
+        final String messageToPrint = new String(data, StandardCharsets.UTF_8);
+
+        log.severe(String.format("Aborting runtime panicked with message: %s", messageToPrint));
```
Comment (stoyan-lime) [inline src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java]: Should this method abort the execution of the runtime, or just log?

Comment (ablax) [reply src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java]: The Polkadot spec doesn't give enough details on what to abort. Also, other hosts haven't used this function, so we might be fine if we remove it.

Comment (ablax) [reply src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java]: Removed in 00d07abeb131e03fe072fb85ce53bac1934a94cf

File: src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java [hunk 3]
```diff
@@ -1,67 +1,151 @@
...
+    }
+
+    /**
+     * Print a number.
+     *
+     * @param number the number to be printed
+     */
+    public void printNumV1(Number number) {
+        log.info("Printing number from runtime: ");
+        System.out.println(number);
```
Comment (georg-getz) [inline src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java]: Why not have this all in the logging line? Do we need to call System here?

File: src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java [hunk 4]
```diff
@@ -1,67 +1,151 @@
...
+        log.log(internalGetLogLevel(level), String.format("Log message from runtime: target=%s, message=%s",
+                targetToPrint,
+                messageToPrint));
     }
 
-    private static Level getLogLevel(int i) {
+    private Level internalGetLogLevel(int i) {
         return switch (i) {
             case 0 -> Level.SEVERE;
             case 1 -> Level.WARNING;
```
Comment (georg-getz) [inline src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java]: Why not leave this as static?

File: src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java (hunk 3 above)
Comment (ablax) [reply src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java]: Not necessarily required to use System output, can add it to the logger's info log.

File: src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java (hunk 4 above)
Comment (ablax) [reply src/main/java/com/limechain/runtime/hostapi/MiscellaneousHostFunctions.java]: It doesn't need to be static.

**Review by georg-getz — APPROVED** [review]

**Pull request merged**